source/server/listener_impl.cc

```diff
@@ -82,100 +82,121 @@ bool shouldBindToPort(const envoy::config::listener::v3::Listener& config) {
 }
 } // namespace
 
-ListenSocketFactoryImpl::ListenSocketFactoryImpl(ListenerComponentFactory& factory,
-                                                 Network::Address::InstanceConstSharedPtr address,
-                                                 Network::Socket::Type socket_type,
-                                                 const Network::Socket::OptionsSharedPtr& options,
-                                                 bool bind_to_port,
-                                                 const std::string& listener_name, bool reuse_port)
+ListenSocketFactoryImpl::ListenSocketFactoryImpl(
+    ListenerComponentFactory& factory, Network::Address::InstanceConstSharedPtr address,
+    Network::Socket::Type socket_type, const Network::Socket::OptionsSharedPtr& options,
+    const std::string& listener_name, uint32_t tcp_backlog_size,
+    ListenerComponentFactory::BindType bind_type, uint32_t num_sockets)
     : factory_(factory), local_address_(address), socket_type_(socket_type), options_(options),
-      bind_to_port_(bind_to_port), listener_name_(listener_name), reuse_port_(reuse_port) {
+      listener_name_(listener_name), tcp_backlog_size_(tcp_backlog_size), bind_type_(bind_type) {
 
-  bool create_socket = false;
   if (local_address_->type() == Network::Address::Type::Ip) {
-    if (socket_type_ == Network::Socket::Type::Datagram) {
-      ASSERT(reuse_port_ == true);
-    }
-
-    if (reuse_port_ == false) {
-      // create a socket which will be used by all worker threads
-      create_socket = true;
-    } else if (local_address_->ip()->port() == 0) {
-      // port is 0, need to create a socket here for reserving a real port number,
-      // then all worker threads should use same port.
-      create_socket = true;
+    if (socket_type == Network::Socket::Type::Datagram) {
+      ASSERT(bind_type_ == ListenerComponentFactory::BindType::ReusePort);
     }
   } else {
     ASSERT(local_address_->type() == Network::Address::Type::Pipe);
     // Listeners with Unix domain socket always use shared socket.
-    create_socket = true;
+    // TODO(mattklein123): This should be blocked at the config parsing layer instead of getting
+    // here and disabling reuse port.
+    if (bind_type_ == ListenerComponentFactory::BindType::ReusePort) {
+      bind_type_ = ListenerComponentFactory::BindType::NoReusePort;
+    }
   }
 
-  if (create_socket) {
-    socket_ = createListenSocketAndApplyOptions();
-  }
+  sockets_.push_back(createListenSocketAndApplyOptions(factory, socket_type, options, bind_type_,
+                                                       listener_name, 0));
 
-  if (socket_ && local_address_->ip() && local_address_->ip()->port() == 0) {
-    local_address_ = socket_->addressProvider().localAddress();
+  if (sockets_[0] != nullptr && local_address_->ip() && local_address_->ip()->port() == 0) {
```

Comment:
See other comments/TODO. The only way this can be null is in the config validation path. I want to get rid of this but this PR is too big as it is. I think I tried to do an early return and it didn't work for some reason, but I can try again.